$OpenBSD: patch-dbus_dbus-connection_c,v 1.8 2011/06/15 16:26:07 ajacoutot Exp $

From FreeBSD:
Linux is more relaxed when it comes to the timeout argument to poll(2).
Linux will accept aby negative number to mean an infinite timeout.
OpenBSD will only accept INFTIM (-1).  Make sure we do not pass an
arbitrary negative number to poll(2) to avoid an infinite EINVAL loop.

--- dbus/dbus-connection.c.orig	Fri Jun 10 21:37:21 2011
+++ dbus/dbus-connection.c	Wed Jun 15 18:17:02 2011
@@ -2354,7 +2354,7 @@ _dbus_connection_block_pending_call (DBusPendingCall *
   DBusConnection *connection;
   dbus_uint32_t client_serial;
   DBusTimeout *timeout;
-  int timeout_milliseconds, elapsed_milliseconds;
+  int timeout_milliseconds, elapsed_milliseconds, remain_milliseconds;
 
   _dbus_assert (pending != NULL);
 
@@ -2435,7 +2435,11 @@ _dbus_connection_block_pending_call (DBusPendingCall *
   _dbus_get_current_time (&tv_sec, &tv_usec);
   elapsed_milliseconds = (tv_sec - start_tv_sec) * 1000 +
 	  (tv_usec - start_tv_usec) / 1000;
-  
+  if (timeout_milliseconds != -1)
+    remain_milliseconds = timeout_milliseconds - elapsed_milliseconds;
+  else
+    remain_milliseconds = -1;
+
   if (!_dbus_connection_get_is_connected_unlocked (connection))
     {
       DBusMessage *error_msg;
@@ -2461,7 +2465,7 @@ _dbus_connection_block_pending_call (DBusPendingCall *
            */
           _dbus_verbose ("dbus_connection_send_with_reply_and_block() waiting for more memory\n");
 
-          _dbus_memory_pause_based_on_timeout (timeout_milliseconds - elapsed_milliseconds);
+          _dbus_memory_pause_based_on_timeout (remain_milliseconds);
         }
       else
         {          
@@ -2470,7 +2474,7 @@ _dbus_connection_block_pending_call (DBusPendingCall *
                                                   pending,
                                                   DBUS_ITERATION_DO_READING |
                                                   DBUS_ITERATION_BLOCK,
-                                                  timeout_milliseconds - elapsed_milliseconds);
+                                                  remain_milliseconds);
         }
 
       goto recheck_status;
@@ -2479,7 +2483,7 @@ _dbus_connection_block_pending_call (DBusPendingCall *
     _dbus_verbose ("dbus_connection_send_with_reply_and_block(): clock set backward\n");
   else if (elapsed_milliseconds < timeout_milliseconds)
     {
-      _dbus_verbose ("dbus_connection_send_with_reply_and_block(): %d milliseconds remain\n", timeout_milliseconds - elapsed_milliseconds);
+      _dbus_verbose ("dbus_connection_send_with_reply_and_block(): %d milliseconds remain\n", remain_milliseconds);
       
       if (status == DBUS_DISPATCH_NEED_MEMORY)
         {
@@ -2489,7 +2493,7 @@ _dbus_connection_block_pending_call (DBusPendingCall *
            */
           _dbus_verbose ("dbus_connection_send_with_reply_and_block() waiting for more memory\n");
 
-          _dbus_memory_pause_based_on_timeout (timeout_milliseconds - elapsed_milliseconds);
+          _dbus_memory_pause_based_on_timeout (remain_milliseconds);
         }
       else
         {          
@@ -2498,7 +2502,7 @@ _dbus_connection_block_pending_call (DBusPendingCall *
                                                   NULL,
                                                   DBUS_ITERATION_DO_READING |
                                                   DBUS_ITERATION_BLOCK,
-                                                  timeout_milliseconds - elapsed_milliseconds);
+                                                  remain_milliseconds);
         }
 
       goto recheck_status;
